daffodil-core/src/main/scala/org/apache/daffodil/core/grammar/ElementBaseGrammarMixin.scala

```diff
@@ -1616,29 +1616,33 @@ trait ElementBaseGrammarMixin
      * the type is a type that respects minLength and maxLength, and the constant length
      * is not in range.
      */
-    val isTypeUsingMinMaxLengthFacets = typeDef.typeNode match {
+    val isTypeUsingLengthOrMinMaxLengthFacets = typeDef.typeNode match {
       case s: NodeInfo.String.Kind => true
       case s: NodeInfo.HexBinary.Kind => true
       case _ => false
     }
     if (
       (lengthKind eq LengthKind.Explicit) &&
-      isTypeUsingMinMaxLengthFacets &&
+      isTypeUsingLengthOrMinMaxLengthFacets &&
       optLengthConstant.isDefined
     ) {
       val len = optLengthConstant.get
-      val maxLengthLong = maxLength.longValueExact
-      val minLengthLong = minLength.longValueExact
+      lazy val maxLengthLong = maxLength.longValueExact
+      lazy val minLengthLong = minLength.longValueExact
+      lazy val lengthLong = length.longValueExact
       def warn(m: String, value: Long): Unit = SDW(
         WarnID.FacetExplicitLengthOutOfRange,
-        "Explicit dfdl:length of %s is out of range for facet %sLength='%s'.",
+        "Explicit dfdl:length of %s is out of range for facet %sength='%s'.",
         len,
         m,
         value,
       )
-      if (maxLengthLong != -1 && len > maxLengthLong) warn("max", maxLengthLong)
-      Assert.invariant(minLengthLong >= 0)
-      if (minLengthLong > 0 && len < minLengthLong) warn("min", minLengthLong)
+      if (hasLength && len != lengthLong) warn("l", lengthLong)
+      else if (hasMinLength || hasMaxLength) {
+        if (maxLengthLong != -1 && len > maxLengthLong) warn("maxL", maxLengthLong)
+        Assert.invariant(minLengthLong >= 0)
+        if (minLengthLong > 0 && len < minLengthLong) warn("minL", minLengthLong)
```

Comment:
Suggest we just pass in the string "minLength", "maxLength" and "length", the above `%sength` is kindof hard to parse and only saves us a few characters.